galaxyproject/galaxy
# [21.05] Do not persist tool state when invoking workflow

If the `tool_inputs` column of the `workflow_step` is loaded as
`MutableJSONType` any "accidental" change of values in the dict marks the
object as "dirty" in the sqlalchemy session it is attached to and will
be persisted when the session is being flushed.
Making the column type `JSONType` instead of `MutableJSONType` means
that the object is not marked as dirty and will not be persisted in that
case. The correct method to use to actually persist step inputs is
WorkflowModule.save_to_step or setting step.tool_inputs directly.  This
fixes https://github.com/galaxyproject/galaxy/issues/9886 and (probably, wasn't able to reproduce myself) https://github.com/galaxyproject/galaxy/issues/12081#issuecomment-860130404

The issue in https://github.com/galaxyproject/galaxy/issues/9886 is that
when you submit a workflow the step state is converted to a tool state
in
https://github.com/mvdbeek/galaxy/blob/90dff9af5bec8bed1258620e5351d0350e284206/lib/galaxy/workflow/modules.py#L661:
```python
    def step_state_to_tool_state(self, state):
        state = safe_loads(state)
        if "format" in state:
            formats = state["format"]
            if formats:
                formats = ",".join(listify(formats))
                state["format"] = formats
        state = json.dumps(state)
        return state
```
and safe_loads just returns the same object if the object that is passed
in is not actually a JSON string. So `state["format"] = formats`
manipulates the step state in place and that gets committed when
flushing the new invocation.


## How to test the changes?
(Select all options that apply)

This is tricky to write good tests for. You could revert https://github.com/galaxyproject/galaxy/pull/9884 and run @dannon's workflow in https://gist.github.com/dannon/d8abf15c848405a7dc15a792149087d3 and make sure that after running it once you don't see `"tool_state": "{\"optional\": false, \"format\": [\"f\", \"\", \"\", \"a\", \"\", \"\", \"s\", \"\", \"\", \"t\", \"\", \"\", \"a\"]}",` in the tool state. That's what I used to figure out how to fix this.

## License
